spring-integration-file/src/main/java/org/springframework/integration/file/remote/RemoteFileTemplate.java

```diff
@@ -434,6 +453,28 @@ public boolean get(final String remotePath, final InputStreamCallback callback)
 	}
 
 	@Override
+	public <T> T invoke(OperationsCallback<F, T> action) {
+		Session<F> contextSession = this.contextSessions.get();
+		if (contextSession == null) {
+			this.contextSessions.set(this.sessionFactory.getSession());
+		}
+		this.activeTemplateCallbacks.incrementAndGet();
```

Comment:
Why? I don't see any problems here with reusing already stored in the `ThreadLocal` session. 
I might agree about your comment in the `finally` block: have had similar doubts. But then decided to leave it somehow... Let me recall what was my conclusion.

We might consider to use additional `ThreadLocal` counter to track the recursion though...